share/p2p/shrexnd/server.go

```diff
@@ -0,0 +1,185 @@
+package shrexnd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/protocol"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/minio/sha256-simd"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	"github.com/celestiaorg/celestia-node/share/getters"
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	pb "github.com/celestiaorg/celestia-node/share/p2p/shrexnd/pb"
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+)
+
+// Server implements server side of shrex/nd protocol to obtain namespaced shares data from remote
+// peers.
+type Server struct {
+	params     *Parameters
+	protocolID protocol.ID
+	getter     share.Getter
+	store      *eds.Store
+	host       host.Host
+
+	cancel context.CancelFunc
+}
+
+// NewServer creates new Server
+func NewServer(host host.Host, store *eds.Store, getter *getters.IPLDGetter, opts ...Option) (*Server, error) {
+	params := DefaultParameters()
+	for _, opt := range opts {
+		opt(params)
+	}
+
+	if err := params.Validate(); err != nil {
+		return nil, fmt.Errorf("shrex-nd: server creation failed: %w", err)
+	}
+
+	srv := &Server{
+		getter:     getter,
+		store:      store,
+		host:       host,
+		params:     params,
+		protocolID: protocolID(params.protocolSuffix),
+	}
+
+	return srv, nil
+}
+
+// Start starts the server
+func (srv *Server) Start(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	srv.cancel = cancel
```

Comment:
This should be a new context because the context given at the start is short-lived. It's a hug because after ten second(default start timeout) no more headers will be served as it's context deadline error

I like though the way you pass it to the stream handler, though cancel still has to be stored(